src/libraries/Microsoft.PowerFx.Interpreter/Functions/IRandomService.cs

```diff
@@ -31,12 +39,25 @@ internal class DefaultRandomService : IRandomService
         [ThreadSafeProtectedByLock(nameof(_randomizerLock))]
         private static readonly Random _random = new Random();
 
+        private static readonly object _randomizerDecimalLock = new object();
+
+        [ThreadSafeProtectedByLock(nameof(_randomizerDecimalLock))]
+        private static readonly Random _randomDecimal = new Random();
+
         public virtual double NextDouble()
         {
             lock (_randomizerLock)
             {
                 return _random.NextDouble();
             }
         }
+
+        public virtual decimal NextDecimal()
+        {
+            lock (_randomizerDecimalLock)
+            {
+                return new decimal(_randomDecimal.Next(), _randomDecimal.Next(), _randomDecimal.Next(542101087), false, 28);
```

Comment:
>_randomDecimal.Next() [](http://example.com/codeflow?start=35&length=21)

This only creates numbers with the lower 31 bits possibly set (the higher bit is always 0), which makes this distribution not uniform.

Reply:
I tried to update logic to make it more random. Please review my new change.

Reply:
Thanks for offline chat and suggestion. I updated logic here.